### PR TITLE
Example of Alpaka-based EventSetup data product that uses PortableCollection member variables

### LINKED ISSUE #721: Create an example of Alpaka ESProduct that has two PortableCollection members

### PR DESCRIPTION
**Pull request by makortel**
#### PR description:

This PR adds an example (as a test) of a case where an Alpaka-based EventSetup data product is defined as a template over the Alpaka device type, and uses `PortableCollection` member variables. More specifically it demonstrates a case where one `PortableCollection` member holds an index to an element of the other `PortableCollection`, and this indirection is hidden as an implementation detail for the user of the EventSetup data product (this is more or less the case that @kakwok has with one of the HCAL EventSetup data products).

In addition, the `TestAlgo.dev.cc` is updated to use the `once_per_grid()` helper.

Resolves https://github.com/makortel/framework/issues/721

#### PR validation:

Unit tests run on CPU and on GPU.

### CODE REVIEW

**Comment by makortel** [comment]
enable GPU

**Comment by makortel** [comment]
@cmsbuild, please test

File: HeterogeneousCore/AlpakaTest/plugins/alpaka/TestAlpakaESProducerE.cc [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+#include <cmath>
+
+namespace ALPAKA_ACCELERATOR_NAMESPACE {
+  /**
+   * This class demonstrates an ESProducer that uses the
+   * PortableCollection-based data model, and that consumes a standard
+   * host ESProduct and converts the data into PortableCollection, and
+   * implicitly transfers the data product to device
+   */
```
Comment (fwyzard) [inline HeterogeneousCore/AlpakaTest/plugins/alpaka/TestAlpakaESProducerE.cc:20]: @makortel can you remind me why an ESProducer that uses this approach needs to be in the `ALPAKA_ACCELERATOR_NAMESPACE` namespace ?
Aren't all the operations happening only on the host, without any reference to the platform, device, or queues ?

**Comment by fwyzard** [comment]
@cmsbuild, please test

File: HeterogeneousCore/AlpakaTest/plugins/alpaka/TestAlpakaESProducerE.cc (hunk 1 above)
Comment (makortel) [reply HeterogeneousCore/AlpakaTest/plugins/alpaka/TestAlpakaESProducerE.cc:20]: > Aren't all the operations happening only on the host, without any reference to the platform, device, or queues ?

The ESProducer indeed _looks_ like that. But under the hood, the component to copy the host-side data product to the devices of the backend is added to the system in the `ALPAKA_ACCELERATOR_NAMESPACE::ESProducer` base class.

Comment (fwyzard) [reply HeterogeneousCore/AlpakaTest/plugins/alpaka/TestAlpakaESProducerE.cc:20]: Mhm... OK.

So if we need to make the same ESProduct available on multiple back-ends, we need to have multiple ESProducers, and run the CPU part multiple times ?

Comment (makortel) [reply HeterogeneousCore/AlpakaTest/plugins/alpaka/TestAlpakaESProducerE.cc:20]: Yes and no. The configuration using more than one non-host backends indeed needs to have an ESProducer for each non-host backend, but out of the box the behavior would be an exception about two ESProducers producing the same host-side ESProduct. The exception can be worked around with an `ESPrefer`, that tells the framework what ESProducer to use the product. So in the end the CPU part is run only once.

Improving the situation towards making this kind of ESProducer truly independent of the backends is in my long term plan (and mentioned in https://github.com/cms-sw/cmssw/issues/40140).

**Comment by fwyzard** [comment]
+heterogeneous

**Comment by rappoccio** [comment]
+1

**Pull request merged**